### PR TITLE
change checks for std: smart pointers

### PR DESCRIPTION
**Pull request by robbietuk**
Addition of compile tests to the CMakeLists.txt to check if unique_ptr exists. See Issue #315

### CODE REVIEW

**Comment by KrisThielemans** [comment]
looks good. let's see what Travis and Appveyor say. However, it now makes sense to do the same for `std::shared_ptr`. Could you give that a go? (One dependency less on Boost.!)

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -181,6 +181,30 @@ include(CheckCXXSymbolExists)
...
+   "
+        STIR_HAVE_UNIQUE_PTR)
+
+if(STIR_HAVE_UNIQUE_PTR)
+    option(STIR_NO_UNIQUE_PTR "Disable use of std::unique_ptr" OFF)
+else()
+    set(STIR_NO_UNIQUE_PTR ON)
+endif()
+
+# Perform test to see if BOOST shared_ptr exists
```
Comment (KrisThielemans) [inline src/CMakeLists.txt]:
```suggestion
# Perform test to see if std::shared_ptr exists
```

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -181,6 +181,30 @@ include(CheckCXXSymbolExists)
...
+check_cxx_source_compiles("
+        #include <memory>
+        std::shared_ptr<int> f(){int *ptr = new int; return std::shared_ptr<int>(ptr);}
+        int main(){auto p = f(); return 0;}
+    "
+        STIR_USE_BOOST_SHARED_PTR
```
Comment (KrisThielemans) [inline src/CMakeLists.txt]:
```suggestion
       HAVE_STD_SHARED_PTR
```

File: src/CMakeLists.txt [hunk 3]
```diff
@@ -181,6 +181,30 @@ include(CheckCXXSymbolExists)
...
+        )
+
 # Now we introduce 2 options such that the user can override the default behaviour
 # (not recommended! unless the above failed of course)
 option(STIR_NO_UNIQUE_PTR
```
Comment (KrisThielemans) [inline src/CMakeLists.txt]: merge this with above code

Comment (KrisThielemans) [reply src/CMakeLists.txt]: also need to fix STIR_USE_BOOST_SHARED_PTR option below in a similar way as the unique_ptr stuff above.

finally need to move the `mark_as_advanced` statement inside the `if`

Comment (robbietuk) [reply src/CMakeLists.txt]: I am not sure if the mark_as_advanced stuff is implemented correctly.

**Comment by KrisThielemans** [comment]
Travis is happy on OSX but I'm afraid all Linux builds are failing. See e.g. https://travis-ci.org/UCL/STIR/jobs/496969217

Can you try this on the VM and see what's happening? Sorry.

**Comment by robbietuk** [comment]
Tested on the VM. It appears that both our new checks (the`check_cxx_source_compiles`) are False so indicating that unique_ptr and std::shared_ptr aren't present.

**Comment by KrisThielemans** [comment]
Please check CMakeFiles/CMakeError.log. Hopefully it tells you something. Feel free to post here.

**Comment by KrisThielemans** [comment]
let's add an explicit C++11 case to Travis:
```
 - os: linux
   python: 3
   env: EXTRA_BUILD_FLAGS="-DDISABLE_CERN_ROOT=0 -DSTIR_OPENMP:BOOL=ON" -DCMAKE_CXX_STANDARD=11 CC=gcc-5 CXX=g++-5

**Comment by robbietuk** [comment]
Compiles for me on a VM. Waiting for checks.

**Pull request merged**